Firestore/core/src/firebase/firestore/remote/remote_objc_bridge.mm

```diff
@@ -127,170 +114,232 @@ bool IsLoggingEnabled() {
-    [mutationProtos addObject:[serializer_ encodedMutation:mutation]];
+  Message<google_firestore_v1_CommitRequest> result{
+      google_firestore_v1_CommitRequest_fields};
+  auto& request = result.mutable_proto();
+
+  request.database = serializer_.EncodeDatabaseId();
+
+  if (!mutations.empty()) {
+    request.writes_count = nanopb::CheckedSize(mutations.size());
+    request.writes = MakeArray<google_firestore_v1_Write>(request.writes_count);
+    pb_size_t i = 0;
+    for (const Mutation& mutation : mutations) {
+      request.writes[i] = serializer_.EncodeMutation(mutation);
+      ++i;
+    }
   }
-  request.writesArray = mutationProtos;
-
-  return request;
-}
 
-grpc::ByteBuffer DatastoreSerializer::ToByteBuffer(GCFSCommitRequest* request) {
-  return ConvertToByteBuffer([request data]);
+  return result;
 }
 
-GCFSBatchGetDocumentsRequest* DatastoreSerializer::CreateLookupRequest(
+Message<google_firestore_v1_BatchGetDocumentsRequest>
+DatastoreSerializer::CreateLookupRequest(
     const std::vector<DocumentKey>& keys) const {
-  GCFSBatchGetDocumentsRequest* request =
-      [GCFSBatchGetDocumentsRequest message];
-
-  request.database = [serializer_ encodedDatabaseID];
-  for (const DocumentKey& key : keys) {
-    [request.documentsArray addObject:[serializer_ encodedDocumentKey:key]];
+  Message<google_firestore_v1_BatchGetDocumentsRequest> result{
+      google_firestore_v1_BatchGetDocumentsRequest_fields};
+  auto& request = result.mutable_proto();
+
+  request.database = serializer_.EncodeDatabaseId();
+  if (!keys.empty()) {
+    request.documents_count = nanopb::CheckedSize(keys.size());
+    request.documents = MakeArray<pb_bytes_array_t*>(request.documents_count);
+    pb_size_t i = 0;
+    for (const DocumentKey& key : keys) {
+      request.documents[i] = serializer_.EncodeKey(key);
+      ++i;
+    }
   }
 
-  return request;
+  return result;
 }
 
-grpc::ByteBuffer DatastoreSerializer::ToByteBuffer(
-    GCFSBatchGetDocumentsRequest* request) {
-  return ConvertToByteBuffer([request data]);
-}
-
-std::vector<MaybeDocument> DatastoreSerializer::MergeLookupResponses(
-    const std::vector<grpc::ByteBuffer>& responses, Status* out_status) const {
+StatusOr<std::vector<model::MaybeDocument>>
+DatastoreSerializer::MergeLookupResponses(
+    const std::vector<grpc::ByteBuffer>& responses) const {
   // Sort by key.
   std::map<DocumentKey, MaybeDocument> results;
 
   for (const auto& response : responses) {
-    auto* proto = ToProto<GCFSBatchGetDocumentsResponse>(response, out_status);
-    if (!out_status->ok()) {
-      return {};
+    auto maybe_proto =
+        Message<google_firestore_v1_BatchGetDocumentsResponse>::Parse(
+            google_firestore_v1_BatchGetDocumentsResponse_fields, response);
+    if (!maybe_proto.ok()) {
+      return maybe_proto.status();
     }
-    MaybeDocument doc = [serializer_ decodedMaybeDocumentFromBatch:proto];
+
+    const auto& proto = maybe_proto.ValueOrDie().proto();
```

Comment:
This pattern makes it such that `DecodeMaybeDocument` can't steal the document message instance--it'll have to make a copy. It's worth thinking about how that could work. Changing this is likely work for a later PR when we actually make that happen, but something to think about until then.

Reply:
Even as we have it today, moving is definitely more efficient than copying because moves are shallow, while copies are deep. Most of the interesting protos have arrays of sub-values making moves of the parent quite cheap because all arrays are dynamically allocated.